distsql/stream.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package distsql
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tipb/go-tipb"
+	goctx "golang.org/x/net/context"
+)
+
+// streamResult implements the SelectResult interface.
+type streamResult struct {
+	resp       kv.Response
+	rowLen     int
+	fieldTypes []*types.FieldType
+	ctx        context.Context
+
+	// NOTE: curr == nil means stream finish, while len(curr.RowsData) == 0 doesn't.
+	curr      *tipb.Chunk
+	scanCount int64
+}
+
+func (r *streamResult) ScanCount() int64 {
+	return r.scanCount
+}
+
+func (r *streamResult) Fetch(goctx.Context) {}
+
+func (r *streamResult) Next(goctx.Context) (PartialResult, error) {
+	var ret streamPartialResult
+	ret.rowLen = r.rowLen
+	finished, err := r.readDataFromResponse(r.resp, &ret.Chunk)
+	if err != nil || finished {
+		return nil, errors.Trace(err)
+	}
+	return &ret, nil
+}
+
+func (r *streamResult) NextChunk(goCtx goctx.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	for chk.NumRows() < r.ctx.GetSessionVars().MaxChunkSize {
```

Comment:
Make `MaxChunkSIze` a field of `streamResult` is more convenient.

Reply:
I make a temporary variable for it. @coocood